### ISSUE TITLE
Ventura stuck or restart in AMD rayzen cpu?

### ISSUE THREAD

**Issue by Mustafa-Agha** [quickemu-project/quickemu#1273]
![01](https://github.com/quickemu-project/quickemu/assets/72730763/c4c44021-5848-494d-b14e-261daeb0a0f5)

It always stuck, and restart, I don't know why?

**Comment by lj3954** [comment]
Please try the latest version of quickemu from git. macOS CPU flags have been changed considerably since v4.9.4 was released.

**Comment by Mustafa-Agha** [comment]
I'm using the latest version 

https://github.com/quickemu-project/quickemu/releases/tag/4.9.4

I used this one

**Comment by lj3954** [comment]
> I'm using the latest version
> 
> https://github.com/quickemu-project/quickemu/releases/tag/4.9.4
> 
> I used this one
> 
> Is this the latest version?
> 
> It installs mojve os correctly, however ventura and somoma don't work.

That is the latest tagged release, but it does not contain the latest changes. While v4.9.5 is not released, please clone the repository and try using the quickemu script directly from it.

**Comment by Mustafa-Agha** [comment]
The same problem happened, I used this repo

git clone --recurse-submodules https://github.com/quickemu-project/quickemu.git

and then used this command

./quickemu --version >> 4.9.5

./quickget macos ventura
./quickemu --vm macos-ventura.conf

first the instance restarted then when I entered (macOS Base System) it became stuck again.

I don't know if this a problem with amd cpu or what?

**Comment by tomchiverton** [comment]
Mac OS Sonoma boots to a black screen here, post installer, having picked the installer once after reboot. This is using quickget/quickemu from Git this morning.

I think something is badly wrong, but the log file is empty.

**Comment by Mustafa-Agha** [comment]
Same here. The log file is empty, and the boot screen is stuck. It won't load the installer.

Tested on ventura and sonima both didn't work, testing it now on monetary to see if this is related to ventura and Sonoma only

**Comment by lj3954** [comment]
@TuxVinyards More needs to be looked into regarding the CPU flag. I think I found something that may be of high importance. I'm not sure what the case is on Intel platforms, but on my AMD system, invtsc, which is a flag that OSX-KVM included in the penryn argument, is reported as constant_tsc instead. Quickemu bash does not catch it. 

For the people here experiencing these issues, I'd recommend trying to launch your VM with my project, [quickemu-rs](https://github.com/lj3954/quickemu-rs). It makes use of a library which interacts directly with the CPU, so it shouldn't have any problems with inaccurate parsing of supported CPU flags. That would help us understand where the issue lies.

**Comment by TuxVinyards** [comment]
@lj3954  I did say that the formula was wrong. I rewrote it and showed you the improvements.

But all I got was a flea in my from @flexiondotorg for "self promotion" of qqX.

Meanwhile, you have become 'The anointed one' and have decided to self promote your ideas  😂

If I had stuck with your ideas, we would all still be working with failure. All you have ever done is moan and complain and say nothing should be changed from what we had back in 4.9.2. 

It's all getting very tedious. 

At least my formula works for me and none of the qqX users are lamenting to me either ....

**Comment by lj3954** [comment]
> @lj3954 I did say that the formula was wrong. I rewrote it and showed you the improvements.
> 
> But all I got was a flea in my from @flexiondotorg for "self promotion" of qqX.
> 
> Meanwhile, you have become 'The anointed one' and have decided to self promote your ideas 😂
> 
> If I had stuck with your ideas, we would all still be working with failure. All you have ever done is moan and complain and say nothing should be changed from what we had back in 4.9.2.
> 
> It's all getting very tedious.
> 
> At least my formula works for me and none of the qqX users are lamenting to me either ....

Rather than addressing what I brought up, you choose to attack me for "promoting my ideas" by recommending testing with a project which should theoretically function in exactly the same way, to see where the issue lies. It's quite strange how you always complain about how unwilling I am to work with others, when virtually none of your work is contributed upstream. Contrast that with my continuation to fixing and improving bash quickemu, despite writing a project intended to supersede it. 

I made it very clear that my position was that no changes should be made unless it was clearly shown that the old approach had issues, and that said issues could be fixed. I don't believe you, me, or anyone else in those discussions had real, tangible issues with the original argument. Those who did are coming to report that one solution, apparently not the one you prefer the most, is also causing issues. Why don't you propose a fix, if you have this magical solution that everyone needs. Unlike you, I'm not recommending my project as a permanent solution for everyone to switch to and completely disregarding everything else, rather, just a step in troubleshooting an issue here.

**Comment by TuxVinyards** [comment]
> I think I found something that may be of high importance. I'm not sure what the case is on Intel platforms, but on my AMD system, invtsc ....

I have already mentioned similar [here](https://github.com/quickemu-project/quickemu/issues/1114#issuecomment-2124719532) 3 weeks ago.

> I made it very clear that my position was that no changes should be made ...

Yes, you did. Continually.

> .... unless it was clearly shown that the old approach had issues

Yes. I did show you.  But you kept ignoring me.

> Why don't you propose a fix, if you have this magical solution that everyone needs.

Not saying it's 'magical' but I did propose one and that got ignored too:  <https://github.com/quickemu-project/quickemu/issues/1114#issuecomment-2129370687>

I then proposed a fix on another issue [here](https://github.com/quickemu-project/quickemu/discussions/1248#discussioncomment-9544913) and got the flea in my ear that I mentioned.

### So, I gave up 
and put my work [here](https://github.com/TuxVinyards/qqX/wiki/Mac-OS#cpu) so people can choose for themselves:

You always have to have areas where you _agree to disagree and find option solutions_ for different ideas.

### But
I can't have entirely given up, otherwise I wouldn't be taking the time to reply, I suppose?

Even after @zen0bit following his myriad of contributions seems to have walked away completely:

I would like to think that it is just hurried people who are not understanding or who are misunderstanding. People who are possibly pressured making snap decisions and comments that they possibly regret but hope people understand as such. And so forth.

I would like to think something might be done about issue #1113 too.  Closed as 'completed' but never fully addressed.

### And
This Quickemu branch [here](https://github.com/TuxVinyards/quickemu/archive/refs/heads/freespirit-dev-tweaks.zip) has the new qqX cpu formula which should be able to run as plain Quickemu without the need to install qqX , if you want to.

**Comment by Mustafa-Agha** [comment]
so you still working on the problem guys, or it was solved?

**Comment by TuxVinyards** [comment]
> so you still working on the problem guys, or it was solved?

You tell us ...

**Comment by lj3954** [comment]
@Mustafa-Agha please try quickemu from https://github.com/TuxVinyards/quickemu/tree/freespirit-dev-tweaks and report back whether it fixes the issue.

**Comment by lj3954** [comment]
@TuxVinyards the comments you made in https://github.com/TuxVinyards/quickemu/commit/174232ceb244dd883b27b811a4893bd1fe4cb777 aren't clear. Are 8 cores required for the VM to boot, or just for macOS to display a processor name?

**Comment by TuxVinyards** [comment]
@lj3954  

> The comments you made in [TuxVinyards@174232c](https://github.com/TuxVinyards/quickemu/commit/174232ceb244dd883b27b811a4893bd1fe4cb777) aren't clear. Are 8 cores required for the VM to boot, or just for macOS to display a processor name?

Presume this comment in the code:

> Tests suggest edit of .conf file to cpu_cores="8"  ie 4 core x 2 threads for MacOS to recognise Xeon  

My guess is that if MacOS  recognises the iMacPro 2017 when it is set at ="8" then things under-the-hood will generally be happier.   

<https://en.wikipedia.org/wiki/IMac_Pro#Technical_specifications>

The OS will boot and run when set at ten or more.  Please feel to experiment but as stated in my [qqX wiki](https://github.com/TuxVinyards/qqX/wiki/Mac-OS)  do backup copy _the whole_  of Mac OS folders before starting. Lots of ancillary files can get changed during processes.

I think the way forward for more powerful setups will be to use Cascade Lake.   I want to have a look at that.

A lot of people, me included, are getting tempted by the new Zen 5 processors.   When I bought my current main processor, I was lucky to even get that.  Chips were really hard to get hold of, certainly here in Europe.  Now we are in the era of desktop CPU's with 16 cores, huge government support for new fabs and 2, 3 and 4 nm becoming common place.

Cascade Lake, the Mac Pro 2019,  can be set to have up to 28 cores, so I am seeing something like `--macpro-2019`  as a  Quickemu switch that would allow users with newer hardware to enable this instead of  a  default of  iMac Pro 2017.  Although, as with Skylake, a lower core number may be needed here too, say 16 ??

<https://en.wikipedia.org/wiki/Mac_Pro#Specifications_3>

The other thing I want to do is to have a config setting that allows individuals to turn off any Qemu CPU flags when an unexpected compatibility problem happens with the physical hardware.  Something in the .conf file like `MacCPU_NoFlags=("pqr" "xy-z")`

**Comment by lj3954** [comment]
> My guess is that if MacOS recognizes the iMacPro 2017 when it is set at ="8" then things under-the-hood will generally be happier

Keep in mind that the actual computer has an 8-core processor with 16 threads. Quickemu, when set to cpu_cores=8 on a host supporting SMT, will emulate a 4-core processor with 8 threads. Also, are you sure the VM will boot with 28 cores selected? I tried it, and it hung immediately. #865 is relevant to this. 

> so I am seeing something like `--macpro-2019` as a Quickemu switch

Good idea, quickemu definitely should have a way to configure the hardware passed to the VM. I think it should be set in the config file though, rather than as a flag. It's not an option that people have to modify each time they boot the VM. I've thought about similar for aarch64 VMs with my quickemu-rs project (i.e. emulating SBCs).

> allows individuals to turn off Qemu CPU flags when an unexpected compatibility problem happens

I think the majority of quickemu users won't have a clue where to start with an option like this. On top of this, you'll have to check each CPU flag for validity before passing it to QEMU. CPU flags should be done out of the box for the user, I think the configuration emulating different hardware would remove the need for this.

**Comment by TuxVinyards** [comment]
> Keep in mind that the actual computer has an 8-core processor with 16 threads. 

:rofl:  Very true.   So many Xeon variants  and _two_ meanings for _core_  just to get everyone confused.  Have I accidentally discovered something?

Here it says ' 8 core'  <https://en.wikipedia.org/wiki/IMac_Pro#Technical_specifications>

But here the term does get clarified:  <https://en.wikipedia.org/wiki/Skylake_(microarchitecture)#Workstation_processors>

There are 4 core (8 thread/core) Xeons  but iMac Pro uses the W-2140B of course.  

So, remembering that we are booting with a virtual CPU, I tested running 8/16 on top of my 6/12 Host, again, so I could get some more screenshots.   Maybe when the About dialog reports Xeon at 4/8  in fact Apple are highlighting the CPU as _not_ being an iMac ....  The other way round.

Performance wise, there was no difference whether I booted  with `cpu_cores=8` or `=16`.

![sonoma-16-core](https://github.com/quickemu-project/quickemu/assets/3956806/0dcb3ea8-e1c8-4160-9ac9-09e3594989e2)

**Comment by lj3954** [comment]
> Edit: perhaps this is the real reason why the above =16 test didn't report Xeon? It was actually running at 6/12 ?

The VM won't boot when set to 6c/12t. You're definitely not being limited to that.

**Comment by TuxVinyards** [comment]
@lj3954 thank you for that.

I have now added `cpu_cores=8 =16 =32` to notes as the being only workable values.  All these three worked for me, even on six physical cores and with no apparent core loss either.  Despite the KVM advisory even 16/32 core machines seemed stable and the requested core numbers were reported as such by lscpu.  How many cores were actually being run by KVM is another matter of course.

The other thing that I have done is to add in optional dev code for CascadeLake to <https://github.com/TuxVinyards/quickemu/tree/freespirit-dev-tweaks> for anyone who wants to try it.

After extracting and diffing the Sky and Cascade code from the Qemu CPU source, I was quite surprised to see how similar the two versions are.  Mac Pro 2019 may sound newer but has only two extra flags.

I was pleased to see `sysctl -a` consistently reporting the correct CPU name, although it's a slight puzzle that even when using Cascade to download and install from scratch, the 'About' dialog still doesn't  report anything other than iMacPro 2017.

Until the 'About' box issue gets solved, being able to run 16c/32t  on Skylake is probably sufficient though and probably the one to go with as it reports more accurately.

Anybody wanting to test either version of this code can download [here](https://github.com/TuxVinyards/quickemu/archive/refs/heads/freespirit-dev-tweaks.zip)  and simply extract to a test folder. Follow up by placing a copy of an existing mac-folder and .conf in with all the files.  Open a terminal in that folder from your file manager and type `./quickemu --vm macos-whatevername`  at the prompt. Note the `./` prefix.

**Comment by lj3954** [comment]
> Fixed via #1208

No, it isn't. Check comment https://github.com/quickemu-project/quickemu/issues/1273#issuecomment-2160616895.

**Comment by flexiondotorg** [comment]
I have tested the tip of git (effectively 4.9.5 release candidate) on several different models of Ryzen CPU and can not reproduce this issue :disappointed: I will test again later using my AMD Ryzen 7 PRO 6850U laptop.

- On the Ryzen CPUs I have available, macOS operation (all Quickemu-supported versions) is faster than it has ever been using 4.9.5.
- The CPU and CPU flag configuration Quickemu sets up in 4.9.5 is based on what my actual x86 Apple Macs report.
- Quickemu 4.9.5 automatically clamps guest CPU cores to a number half that of the host.
  - Then, for macOS guests, the derived CPU core count is limited to the nearest but lowest power of 2 using a predefined table to ensure core configurations documented by the [OSX-KVM](https://github.com/kholia/OSX-KVM) project as problematic are avoided.
- The `invtsc` CPU flag is passed through automatically by Quickemu.
- The `constant_tsc` CPU flag can not be passed to the Haswell-v2 and Skylake-Server-v3 CPU models offered by QEMU:

```
qemu-system-x86_64: Property 'Skylake-Server-v3-x86_64-cpu.constant-tsc' not found
```

I'll leave this open to gather more information that can hopefully identify the root cause. I welcome pull requests that demonstrably address whatever the root cause of this issue might be.

**Comment by popey** [comment]
I may have an option to move forward from a wedged / black screen macOS VM, part way through the install - likely stuck in the EFI with the screen blanked.

Using quickemu 4.9.5 from the PPA on Ubuntu 24.04 on my ThinkPad Z13. My system:

![image](https://github.com/quickemu-project/quickemu/assets/1841272/74c9b2b6-b348-49a0-872e-a7fb69e680fe)

I tried installing macOS Ventura. I walked away to do some work, and when I came back, I had a black screen, thus:

![image](https://github.com/quickemu-project/quickemu/assets/1841272/2618d3fe-54cb-40de-89e5-480c274dd560)

I tried to anything in the window (click with mouse, press keys) nothing happened. I couldn't determine the state.

However, I was able to use the Qemu [monitor](https://qemu-project.gitlab.io/qemu/system/monitor.html) to inject keypresses that 'woke' the screen up.

You'll need 'socat' to do this, and know the path to the `macos-ventura-monitor.socket` (or similarly named file) in the quickemu macos VM folder. Here's what I used:

`$ socat -  unix-connect:/home/alan/VMs/macos-ventura/macos-ventura-monitor.socket`

You'll get this prompt from QEMU:

```
QEMU 8.2.2 monitor - type 'help' for more information
(qemu) 
```

You can send keypresses, and mouse clicks. So at the `(qemu)` prompt, type `sendkey` then the key you want to send. There's a prescribed named list of keys.

* `ret` = Return
* `esc` = Escape
* `left` = Left arrow
* `right` = Right arrow
* `spc` = Space bar

So, `sendkey esc` which 'virtually presses' the escape key.

```
(qemu) sendkey esc
sendkey esc
```

etc.

You can also click the mouse with `mouse_button n` where `n` is the mouse button number 1, 2, 3 etc.

```
(qemu) sendkey mouse_button 1
mouse_button 1
```

This woke the display up. It's possible I may have pressed a number of keys (left, right, ret etc) so worth a play.

![image](https://github.com/quickemu-project/quickemu/assets/1841272/08708f03-d86c-44f6-8cb2-65304a11b603)

I was then able to progress the install:

![image](https://github.com/quickemu-project/quickemu/assets/1841272/91c942b7-27da-4b02-adee-4b43ded1b42f)

Maybe give that a try?

**Comment by lj3954** [comment]
> can not reproduce this issue

Neither can any of us. macOS issues are extremely hardware dependent, we've been waiting for @Mustafa-Agha to try @TuxVinyards' quickemu branch which has a different way of handling macOS CPU arguments. 

> The constant_tsc CPU flag can not be passed to the Haswell-v2 and Skylake-Server-v3 CPU models offered by QEMU:

To the best of my knowledge, constant_tsc and invtsc are the same thing. It's just another instance of QEMU using a different name for the CPU flag than lscpu. I think it's also probable that the Haswell and especially Skylake CPUs have this flag enabled by default, so it's nothing to worry about.

**Comment by flexiondotorg** [comment]
It seems that everyone having stalls/lockups running newer macOS with Quickemu is using a mobile Ryzen SKU of 4000U or 5000U series. 

I have tested on two Ryzen 6000U SKUs and can not reproduce the issue. So, I think we have an avenue of investigation. I don't have a 4000U/5000U system to test with sadly :disappointed:

**Comment by tomchiverton** [comment]
Maybe just some more detailed logging would help - possible CPU instructions used, ones used after filter etc ?

**Comment by puresick** [comment]
> It seems that everyone having stalls/lockups running newer macOS with Quickemu is using a mobile Ryzen SKU of 4000U or 5000U series.
> 
> I have tested on two Ryzen 6000U SKUs and can not reproduce the issue. So, I think we have an avenue of investigation. I don't have a 4000U/5000U system to test with sadly 😞

@flexiondotorg I can confirm that I have similar issues on a AMD Ryzen™ 7 PRO 5850U.

**Comment by br0kenpixel** [comment]
Are there any updates on this issue? I can confirm that I'm also unable to boot anything newer than Big Sur on 7945HX. Interestingly, if I try this under an Ubuntu VM (with nested virtualisation), it works.

Ubuntu 24.04 however uses an older version of QEMU (8.2.2), while Fedora 41 uses 9.1.2. I tried compiling 8.2.8 on Fedora and it has the same issue. I don't know what's going on.

**Comment by damiandudycz** [comment]
Any updates on this? I was able to install macOS Sequoia, but when running the installed system, it just freezes with hight cpu usage. Same with other macOS versions. Im running AMD Ryzen Z1 Extreme (Zen 4) inside Windows 11 SDL 2.

**Comment by warnerbryce** [comment]
I have the same problem Ryzen 7840U.
I have a working machine under VirtualBox.

**Comment by jimdigriz** [comment]
The issue here for me is the: `panic(cpu 0 caller 0xfff...): Non-monotonic time`

<img width="1685" height="1079" alt="image" src="https://github.com/user-attachments/assets/8bc4bdf1-a4b9-45a7-8c81-e731e0de289d" />

From reading around, this is due to the TSC being broken for my Ryzen (apparently common on older Ryzen CPUs), so on the host you may see something like this:
```
alex@hanzawa:~$ lscpu 
Architecture:                x86_64
  CPU op-mode(s):            32-bit, 64-bit
  Address sizes:             48 bits physical, 48 bits virtual
  Byte Order:                Little Endian
CPU(s):                      16
  On-line CPU(s) list:       0-15
Vendor ID:                   AuthenticAMD
  Model name:                AMD Ryzen 7 5700U with Radeon Graphics
    CPU family:              23
    Model:                   104
    Thread(s) per core:      2
    Core(s) per socket:      8
    Socket(s):               1
    Stepping:                1
    Frequency boost:         enabled
    CPU(s) scaling MHz:      25%
    CPU max MHz:             4372.0000
    CPU min MHz:             400.0000
    BogoMIPS:                3593.23
    Flags:                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep
                             _good nopl xtopology nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand lahf_
                             lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_
                             l3 cdp_l3 hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 cqm_llc
                              cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd cppc arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeas
                             sists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca
Virtualization features:     
  Virtualization:            AMD-V
Caches (sum of all):         
  L1d:                       256 KiB (8 instances)
  L1i:                       256 KiB (8 instances)
  L2:                        4 MiB (8 instances)
  L3:                        8 MiB (2 instances)
NUMA:                        
  NUMA node(s):              1
  NUMA node0 CPU(s):         0-15
Vulnerabilities:             
  Gather data sampling:      Not affected
  Indirect target selection: Not affected
  Itlb multihit:             Not affected
  L1tf:                      Not affected
  Mds:                       Not affected
  Meltdown:                  Not affected
  Mmio stale data:           Not affected
  Reg file data sampling:    Not affected
  Retbleed:                  Mitigation; untrained return thunk; SMT enabled with STIBP protection
  Spec rstack overflow:      Mitigation; Safe RET
  Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
  Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
  Spectre v2:                Mitigation; Retpolines; IBPB conditional; STIBP always-on; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
  Srbds:                     Not affected
  Tsa:                       Not affected
  Tsx async abort:           Not affected
  Vmscape:                   Mitigation; IBPB before exit to userspace

alex@hanzawa:~$ sudo dmesg | grep -i tsc
[sudo] password for alex: 
[    0.000000] tsc: Fast TSC calibration using PIT
[    0.000000] tsc: Detected 1796.617 MHz processor
[    0.077280] clocksource: tsc-early: mask: 0xffffffffffffffff max_cycles: 0x19e5ade6ca6, max_idle_ns: 440795284310 ns
[    0.347433] clocksource: Switched to clocksource tsc-early
[    1.624276] tsc: Refined TSC clocksource calibration: 1810.296 MHz
[    1.624294] clocksource: tsc: mask: 0xffffffffffffffff max_cycles: 0x1a1827b323d, max_idle_ns: 440795216569 ns
[    1.624602] clocksource: Switched to clocksource tsc
[    2.584283] clocksource: timekeeping watchdog on CPU4: Marking clocksource 'tsc' as unstable because the skew is too large:
[    2.584304] clocksource:                       'tsc' cs_nsec: 479378261 cs_now: 2955fbebf0 cs_last: 29224217b8 mask: ffffffffffffffff
[    2.584310] clocksource:                       Clocksource 'tsc' skewed -3647933 ns (-3 ms) over watchdog 'hpet' interval of 483026194 ns (483 ms)
[    2.584318] clocksource:                       'tsc' is current clocksource.
[    2.584332] tsc: Marking TSC unstable due to clocksource watchdog                     <---------------------------------------------
[    2.584356] TSC found unstable after boot, most likely due to broken BIOS. Use 'tsc=unstable'.
[    2.584931] clocksource: Checking clocksource tsc synchronization from CPU 8 to CPUs 0-1,3,7,14.
[   14.938630] kvm_amd: TSC scaling supported
[  414.503270] kvm: SMP vm created on host with unstable TSC; guest TSC will not be reliable            <------------------------------------------------
```
[quickreport.txt](https://github.com/user-attachments/files/24476445/quickreport.txt)

Apparently this can be [fixed with a BIOS update for some](https://github.com/thenickdude/KVM-Opencore/issues/15#issuecomment-1603398655), but for my ASUSTek MINIPC PN51-E1 is at the latest BIOS version (03/04/2025) so that is probably the end of the road for me.

One workaround is to [use just a single CPU (`--extra_args '-smp cores=1,threads=1,sockets=1'`)](https://github.com/dockur/macos/issues/84)...not exactly ideal.

Another is to somehow keep resync'ing the TSC host side. There was a [proposal for Linux a while back](https://lkml.org/lkml/2008/9/26/35) but it seems to have gone nowhere.

Maybe another fix would be to disable CPU frequency scaling and power saving...

I suspect https://github.com/orgs/quickemu-project/discussions/1274 may also be affected, it just may be that their TSC desync occurs *during* them running the VM, rather than during boot like myself.

**Comment by jimdigriz** [comment]
> Another is to somehow keep resync'ing the TSC host side...

Or it turns out we [can just ignore the problem](https://github.com/RPCS3/rpcs3/issues/16545) and add `tsc=reliable` to our kernel parameters and reboot.

Do not use this option blindly, it should only be used if your TSC warps only during early boot (ie. once and then is stable). If instead it regularly warps, you are likely to have a Bad Day(tm).

Your Mileage May Vary.

**Comment by br0kenpixel** [comment]
Unfortunately my system keeps switching to hpet as soon as I start the VM. Disabling hpet just causes acpi_pm to be used instead. At this point, a BIOS update would probably be the only solution, but I don't see that happening for my Legion Pro.

**Comment by jimdigriz** [comment]
@flexiondotorg PR #1808 did *not* fix this.

**Comment by flexiondotorg** [comment]
See the notes, in that PR. This is the best we can do, it is not a Quickemu bug 😞

**Comment by br0kenpixel** [comment]
> [@flexiondotorg](https://github.com/flexiondotorg) PR [#1808](https://github.com/quickemu-project/quickemu/pull/1808) did _not_ fix this.

What version of macOS are you trying to use? For me, the PR only fixed Catalina and Big Sur. Newer versions still don't work, but also have stricter TSC requirements.

**Comment by jimdigriz** [comment]
> See the notes, in that PR. This is the best we can do, it is not a Quickemu bug

@flexiondotorg Understood, but how is a user going to know to look here when things just do not work?

More so as there is not really anything to search for as the actual 'Non-monotonic' error gets scrolled off the display before the user notices there is a problem.

> What version of macOS are you trying to use?

@br0kenpixel please read my input here on https://github.com/quickemu-project/quickemu/issues/1273#issuecomment-3719810362 and https://github.com/quickemu-project/quickemu/issues/1273#issuecomment-3719864146

I *understand* the bug and the workaround I provided, but PR #1808 does nothing to effect this issue.

The problem all this is not obvious to the user, it seems to be non-obvious to a lot of people out there and is barely mentioned when you do know what to look for. Most just have marked issues in their similar projects with "need more user input", "unable to reproduce" or "user system must be bust as it works on my Ryzen..."

Of course there is no magic you can do to make the TSC reliable, but maybe it would be helpful to emit a message to the user so there is something actionable they can do based on:
 * check `/sys/bus/clocksource/devices/clocksource0/{available,current}_clocksource`
 * check `/proc/cmdline` for `tsc=...`

For me, my TSC is only unstable early on boot, afterwards it seems okay. Fortunately I get to use `tsc=reliable` and maybe others can too so we can avoid telling people to go all out and "get another CPU...".

Maybe this will help with holding back the "does not work" issues that may get opened.

Myself, I have a workaround, I needed to document that here as well as that PR #1808 is not a fix for this.  If only to prevent others getting confused; especially as the `invtsc`<->`constant_tsc` PR makes no difference for me on my Ryzen 7 5700U and the "pretend it *is* stable" does and still works on a version of `quickemu` without that PR applied.

**Comment by tomchiverton** [comment]
Agree. 

Quickemu can spot the CPU, and knows what MacOS version is booting, and if it's a known bad combination it should say so as a warning.

-- 
†øღ

**Comment by flexiondotorg** [comment]
This is a known hardware limitation affecting certain AMD Ryzen mobile CPUs (4000U, 5000U series, and some Zen 4 mobile chips) where the TSC is marked unstable during boot, causing macOS Ventura and newer to freeze.

Documentation for this issue is now available in the wiki: **[TSC instability on AMD Ryzen mobile CPUs](https://github.com/quickemu-project/quickemu/wiki/03-Create-macOS-virtual-machines#tsc-instability-on-amd-ryzen-mobile-cpus)**

**Two options:**
1. Add `tsc=reliable` to your kernel boot parameters (if your TSC stabilises after boot)
2. Use macOS Big Sur or earlier, which are more tolerant of TSC instability

The wiki page includes instructions for checking if you're affected and step-by-step configuration for both GRUB and systemd-boot.

**Comment by br0kenpixel** [comment]
For anyone still looking for a solution, if you're on affected AMD hardware, try the latest CachyOS RC kernel, they recently [added some patches](https://github.com/CachyOS/linux-cachyos/issues/925) to make TSC work.

This may not be a solution for everyone, but it's worth a try. It resolved the issue for me.
